### ISSUE TITLE
Hyperparameter tuning fails for YOLO26 on Google Colab (all iterations fail)

### ISSUE THREAD

**Issue by lenczo** [ultralytics/ultralytics#23200]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component
Hyperparameter Tuning

### Bug

I am training a custom dataset using a YOLO26 model in Google colab. A normal training run with model.train() works without issues, but hyperparameter tuning with model.tune() consistently fails. I am trying to understand what causes the tuning failure and how to fix it.

My goal is to first tune hyperparameters and then train a YOLO26 model on my own dataset.

Dataset Details
- Custom dataset with ~300 images
- 2 classes
- One object per image
- Dataset is fully labeled and validated
- Training works when using model.train()

What I Tried
- Running a single training run first (as suggested in the docs) → works
- Moving dataset between Google Drive and local Colab storage → no change
- Adjusting batch size and workers → no change
- Verifying dataset and paths (data.yaml) → works for training

### Environment

Ultralytics 8.4.0 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 39.8/112.6 GB disk)

### Code
from ultralytics import YOLO

model = YOLO("yolo26s.pt")

model.tune(
    data="/content/dataset/dataset_split/data.yaml",
    epochs=30,
    iterations=200,
    optimizer="AdamW",
    batch=4,
    resume=True,
    workers=2
)

### Error Message
Starting iteration 300/300 with hyperparameters: {'lr0': 0.0095, 'lrf': 0.01112, 'momentum': 0.95035, 'weight_decay': 4e-05, 'warmup_epochs': 2.49153, 'warmup_momentum': 0.95, 'box': 5.87944, 'cls': 0.51261, 'dfl': 1.56955, 'hsv_h': 0.02654, 'hsv_s': 0.9, 'hsv_v': 0.37694, 'degrees': 0.00279, 'translate': 0.11135, 'scale': 0.91767, 'shear': 0.00074, 'perspective': 0.0001, 'flipud': 0.0, 'fliplr': 0.53575, 'bgr': 0.0, 'mosaic': 0.95386, 'mixup': 0.00535, 'cutmix': 0.00326, 'copy_paste': 8e-05, 'close_mosaic': 10} ERROR ❌ training failure for hyperparameter tuning iteration 300 Command '['/usr/bin/python3', '-m', 'ultralytics.cfg.__init__', 'train', 'task=detect', 'mode=train', 'model=yolo26s.pt', 'data=/content/drive/MyDrive/00_MASTERARBEIT/dataset_split/data.yaml', 'epochs=30', 'time=None', 'patience=100', 'batch=16', 'imgsz=640', 'save=True', 'save_period=-1', 'cache=False', 'device=None', 'workers=8', 'project=None', 'name=None', 'exist_ok=False', 'pretrained=True', 'optimizer=AdamW', 'verbose=True', 'seed=0', 'deterministic=True', 'single_cls=False', 'rect=False', 'cos_lr=False', 'close_mosaic=10', 'resume=False', 'amp=True', 'fraction=1.0', 'profile=False', 'freeze=None', 'multi_scale=0.0', 'compile=False', 'overlap_mask=True', 'mask_ratio=4', 'dropout=0.0', 'val=True', 'split=val', 'save_json=False', 'conf=None', 'iou=0.7', 'max_det=300', 'half=False', 'dnn=False', 'plots=True', 'source=None', 'vid_stride=1', 'stream_buffer=False', 'visualize=False', 'augment=False', 'agnostic_nms=False', 'classes=None', 'retina_masks=False', 'embed=None', 'show=False', 'save_frames=False', 'save_txt=False', 'save_conf=False', 'save_crop=False', 'show_labels=True', 'show_conf=True', 'show_boxes=True', 'line_width=None', 'format=torchscript', 'keras=False', 'optimize=False', 'int8=False', 'dynamic=False', 'simplify=True', 'opset=None', 'workspace=None', 'nms=False', 'lr0=0.0095', 'lrf=0.01112', 'momentum=0.95035', 'weight_decay=4e-05', 'warmup_epochs=2.49153', 'warmup_momentum=0.95', 'warmup_bias_lr=0.1', 'box=5.87944', 'cls=0.51261', 'dfl=1.56955', 'pose=12.0', 'kobj=1.0', 'rle=1.0', 'angle=1.0', 'nbs=64', 'hsv_h=0.02654', 'hsv_s=0.9', 'hsv_v=0.37694', 'degrees=0.00279', 'translate=0.11135', 'scale=0.91767', 'shear=0.00074', 'perspective=0.0001', 'flipud=0.0', 'fliplr=0.53575', 'bgr=0.0', 'mosaic=0.95386', 'mixup=0.00535', 'cutmix=0.00326', 'copy_paste=8e-05', 'copy_paste_mode=flip', 'auto_augment=randaugment', 'erasing=0.4', 'cfg=None', 'tracker=botsort.yaml', 'save_dir=/content/runs/detect/train']' returned non-zero exit status 1. WARNING ⚠️ No valid fitness values to plot (all iterations may have failed)

!But the Error occurs at all iteration steps, so its the same for 1/300 till 300/300

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lenczo, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help us triage faster, and an Ultralytics engineer will also assist soon 🙏

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us debug it. In your case, tuning appears to be launching a subprocess and failing with a non-zero exit status, but the log excerpt doesn’t include the underlying traceback from inside that subprocess. An MRE that captures the *full* output would be super helpful ✅

Please include in the MRE:
-Exact Colab notebook steps (from fresh runtime) and the exact commands run
-Your `pip` install step and `ultralytics` version confirmation
-A small sample dataset (or instructions to reproduce with a public dataset) + `data.yaml`
-The complete tuning output including the *first* failure (iteration 1/…); not only the final summary line
-Your `runs/` directory contents for a failed iteration (if any), especially the produced logs

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can try

```
pip install git+https://github.com/ultralytics/ultralytics@tune-fix
```

**Comment by lenczo** [comment]
Thanks for the quick answer. I tried it with `pip install git+https://github.com/ultralytics/ultralytics@tune-fix`, but unfortunatly same Error occurs. 

Here are some additions:

**Pip install step**
`!pip install -U ultralytics
!yolo settings sync=False
import ultralytics
ultralytics.checks()
`
**Output**
`Ultralytics 8.4.0 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 38.7/112.6 GB disk)`

**tuning step**
`from ultralytics import YOLO
model = YOLO("yolo26s.pt")
model.tune(
    data="/content/drive/MyDrive/00_MASTERARBEIT/dataset_split/data.yaml",
    epochs=30,
    iterations=300,
    optimizer="AdamW",
    batch=4,
    resume=True,
    workers=2
)
`
_Since I had already iterated up to 200, I had to increase the epochs to 300 so that it would tune again, so 201 is my first iteration step._

[Tuner_output.txt](https://github.com/user-attachments/files/24614621/Tuner_output.txt)

[tune_results.csv](https://github.com/user-attachments/files/24614643/tune_results.csv)

[best_hyperparameters.yaml](https://github.com/user-attachments/files/24614644/best_hyperparameters.yaml)

**data.yaml**
path: /content/dataset_split/
train: images/train
val: images/val
test: images/test
names:
  0: blocked
  1: clear

**dataset image examples**

![Image](https://github.com/user-attachments/assets/c5fc6137-d894-4a11-ac30-2c8cb5d75796)

**Comment by Y-T-G** [comment]
You shouldn't use `pip install -U ultralytics` after using the command I gave. Reset runtime and only run the command I gave to install Ultralytics

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #23191: Exclude `save_dir` from argument validation](https://github.com/ultralytics/ultralytics/pull/23191).

### What changed (and why it may fix this)
- **`save_dir` is now treated as a supported override/config key**, so it can be set reliably via CLI or Python overrides.
- The previous behavior in `get_cfg()` that **silently dropped `save_dir`** when it wasn’t present in the base config has been removed.
- `check_dict_alignment()` was updated to **allow `save_dir` as a custom key** (similar to `augmentations`), preventing “unknown key” validation errors.

This should make output directory handling more predictable across runs and workflows that manage experiment folders externally.

### Credits
Thanks to **@Y-T-G** for the PR, with contributions from **@glenn-jocher**.

### How to test the fix
- Install from the latest `main` branch:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Alternatively, waiting for the next official PyPI release is also an option if installing from source isn’t ideal.

### Feedback requested
If possible, please retry the original repro steps with the updated code and share whether `save_dir` overrides now behave as expected.

Thanks again 🙏 for reporting this, and any additional details are welcome if anything still seems off.

**Comment by lenczo** [comment]
Hey, 

thanks for the quick bug-fix. Everything works now :)

**Comment by glenn-jocher** [comment]
Good to hear it’s resolved—if anything else crops up (especially in Colab), please open a new issue with `yolo checks` output and the full text traceback so we can reproduce it.

**Comment by Y-T-G** [comment]
You should open a new issue with all the information because your issue is not related to this
